File: 78-Subsets/78-Subsets.java
// Last updated: 01/09/2025, 01:06:06
class Solution {
    public void creating(List<List<Integer>>res,List<Integer>lst, int ind, int size, int[] nums){
        if(ind == size){
            res.add(new ArrayList<>(lst));
            return;
        }

        // pick
        lst.add(nums[ind]);
        creating(res, lst, ind+1, size, nums);
        lst.remove(lst.size() - 1);
        
        // not pick
        creating(res, lst, ind + 1, size, nums);
    }

    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>>res = new ArrayList<>();
        List<Integer>lst = new ArrayList<>();
        creating(res,lst, 0, nums.length , nums);
        return res;
    }
}

File: 78-Subsets/78-Subsets.cpp
// Last updated: 01/09/2025, 01:06:09
class Solution {
public:
    void sub(int i , int n, vector<int>&smallVec, vector<int>&givenVec, vector<vector<int>>&res){
        if(i == n){
            res.push_back(smallVec);
            return;
        }
        
        // pick
        smallVec.push_back(givenVec[i]);
        sub(i+1,n,smallVec,givenVec,res);
        smallVec.pop_back();

        // not pick
        sub(i+1,n,smallVec,givenVec,res);
    }
    vector<vector<int>> subsets(vector<int>& nums) {
        int n = nums.size();
        vector<int>vec;
        vector<vector<int>>res;
        sub(0,n,vec,nums,res);
        return res;
    }
};